### PR TITLE
SI-7614 Minimize the times of evaluation f in TraversableOnce.maxBy/minBy

### PR DESCRIPTION
**Pull request by yllan**
In the previous implementation, maxBy/minBy will evaluate most of its elements
with f twice to get the ordering. That results (2n - 2) evaluations of f.

I save both the element and result of evaluation to a tuple so that it doesn't
need to re-evaluate f on next comparison. Thus only n evaluations of f, that is
the optimal.

### CODE REVIEW

**Comment by soc** [comment]
I suggest picking a while loop/foreach for readability reasons instead.
See https://groups.google.com/d/msg/scala-internals/uAYeQULo0bM/8Od_k-WgUGEJ.

**Comment by yllan** [comment]
Okay. Please check it now.

**Comment by adriaanm** [comment]
Welcome, @yllan! Please sign the [Scala CLA](http://typesafe.com/contribute/cla/scala)

**Comment by yllan** [comment]
OK @adriaanm. I've signed the Scala CLA in the provided link.

**Comment by adriaanm** [comment]
tests in the pos/ category are only compiled, not run -- they are meant to check that something compiles successfully
I think it would be better to use junit here, @gkossakowski can explain how.

**Comment by yllan** [comment]
What is the recommended location suited for this test? test/files/run?

**Comment by gkossakowski** [comment]
@yllan: you can use junit tests directly. They are faster to run that partest tests and fit very well testing library code.  You need to put them into `test/junit` folder. Check 25a8e70c134c339be0f436013828cacb4898cbe3 and 173e709017f5cd489e5526bec92da8dc904fc436 to see how our junit tests work.

**Comment by yllan** [comment]
Thanks for the instruction. I've moved the test to junit.

**Comment by gkossakowski** [comment]
@Ichoran: would like to review this PR?

**Comment by Ichoran** [comment]
Not quite okay, by visual inspection.  Reverses the logic of which element to take on ties (last as implemented in patch, first originally).  Simple fix, though--just use gt instead of gteq in the new code.  I just looked, didn't test, so probably the thing to do is write a unit test that verifies this, and trust the test more than my eyes.

File: src/library/scala/collection/TraversableOnce.scala [hunk 1]
```diff
@@ -221,13 +221,37 @@ trait TraversableOnce[+A] extends Any with GenTraversableOnce[A] {
...
       throw new UnsupportedOperationException("empty.maxBy")
 
-    reduceLeft((x, y) => if (cmp.gteq(f(x), f(y))) x else y)
+    var maxF: B = null.asInstanceOf[B]
+    var maxElem: A = null.asInstanceOf[A]
+    var first = true
+
+    for (elem <- self) {
+      val fx = f(elem)
+      if (first || cmp.gteq(fx, maxF)) {
```
Comment (Ichoran) [inline src/library/scala/collection/TraversableOnce.scala]: gteq here changes logic from returning first match to returning last match.  Use gt or !gteq(maxF, fx) or somesuch.

File: src/library/scala/collection/TraversableOnce.scala [hunk 2]
```diff
@@ -221,13 +221,37 @@ trait TraversableOnce[+A] extends Any with GenTraversableOnce[A] {
...
       throw new UnsupportedOperationException("empty.minBy")
 
-    reduceLeft((x, y) => if (cmp.lteq(f(x), f(y))) x else y)
+    var minF: B = null.asInstanceOf[B]
+    var minElem: A = null.asInstanceOf[A]
+    var first = true
+    
+    for (elem <- self) {
+      val fx = f(elem)
+      if (first || cmp.lteq(fx, minF)) {
```
Comment (Ichoran) [inline src/library/scala/collection/TraversableOnce.scala]: lteq here changes logic from returning first match to last match.  Use lt or !lteq(minF, fx) or somesuch.

File: src/library/scala/collection/TraversableOnce.scala (hunk 1 above)
Comment (yllan) [inline src/library/scala/collection/TraversableOnce.scala]: Good point! Is this behaviour specified in any documentation? 
Maybe I could address this behaviour in document as well.

**Comment by yllan** [comment]
Thanks @Ichoran. I've added this behavior both in scaladoc and junit test.

**Comment by Ichoran** [comment]
@gkossakowski - LGTM with the latest revisions.

**Comment by gkossakowski** [comment]
Excellent! Thanks @yllan and @lchoran!

**Pull request merged**